### ISSUE TITLE
Taxonomy: Hierarchical Focus Trap

### ISSUE THREAD

**Issue by BrittanyIRL** [swissspidy/web-stories-wp#9274]
## Bug Description

The category list should be ignored with tabs, but if we use the arrow keys we should be able to navigate between options. That way we don’t have to tab through all the categories when trying to get to the `Tags` section

While here, let's update the HTML to be more semantic as well. @barklund points out: 

* Use a non-collapsible, multi-selectable [tree view](https://www.w3.org/TR/wai-aria-practices-1.1/#TreeView) for the widget.
  * A `role="tree"` around the whole thing.
  * A `role="group"` around any list of sibling category nodes
  * A `role="treeitem"` around any single category node
* Remember to set `aria-multiselectable="true"` and set `aria-selected` specifically to true or false for the individual selectable elements (if attribute is missing, it means that the element cannot be selected, so actively set to false).
* Note the quite complex list of keyboard interactions, but let's only implement what we strictly need:
  * We can ignore expanding/collapsing nodes, because we only support all nodes expanded for now (I think that's okay).
  * Also ignore the type-ahead instructions, we have a search field.
  * And finally ignore the somewhat complex instructions for multi-selecting while navigating with <kbd>Shift</kbd> or <kbd>Ctrl</kbd>.
  * We'll only support moving up and down using arrow keys (and optionally end/home) as well as toggling selected state with <kbd>Space</kbd> (not <kbd>Enter</kbd> for some reason 🤷 ).
* Also note initial focus should go to the first element if none are selected or the first selected element if any are selected when receiving focus.
  * Optionally, if receiving focus in reverse tab order (shift-tabbing from a subsequent element), the last element if none selected or the last selected element should receive focus. As [this is not actually in the WAI-ARIA recommendations](https://www.w3.org/TR/wai-aria-practices-1.1/#:~:text=when%20a%20multi-select%20tree%20receives%20focus), it's most definitely skippable.
* Refer to the [example implementations](https://www.w3.org/TR/wai-aria-practices-1.1/#examples-13) for further guidance. 

Fixing this here in tandem with the keyboard nav should make things easier.

## Expected Behaviour

Once you get to the categories/hierarchical component that lists all relevant taxonomies, you should use the arrow keys to navigate through the options not the tab key, if tab is hit then it should move focus past the list of terms.

## Additional Context

- Plugin Version:
- WordPress Version:
- Operating System:
- Browser:

**Comment by barklund** [comment]
I actually recommended using listbox because I didn't think a tree view supported selection. [But it does](https://www.w3.org/TR/wai-aria-practices-1.1/#TreeView)!

So I'll update the text to recommend using a tree view instead.

**Comment by BrittanyIRL** [comment]
Thanks! I just copied in all of the bug bash writing as is thinking discovery on the comments would be done in ticket so this is perfect, appreciate the update!

**Comment by samwhale** [comment]
The categories cannot be collapsed, so do we need the `tree` role? All of the examples use collapsible options. ![image.png](https://images.zenhubusercontent.com/5b1032a1348b7966b43421a4/12dc43b1-d4e8-4a01-a935-d3a946324c92)

If that were to change, then `role="tree"` would need to be added. https://www.w3.org/TR/wai-aria-practices-1.1/#TreeView

**Comment by BrittanyIRL** [comment]
So, sounds like this is definitely more of a [listbox](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role). This is also an improvement on the categories list in 'posts' which uses tab.

**Comment by samwhale** [comment]
Yeah that's good - https://www.w3.org/TR/wai-aria-practices-1.1/#Listbox. ~Seems like it doesn't need the `aria-multiselect` either since the role implies the multi-select-ability~ maybe it does still 🤔

**Comment by samwhale** [comment]
Oh it does still - good to know: ![image.png](https://images.zenhubusercontent.com/5b1032a1348b7966b43421a4/65f14311-155d-4dc6-b409-ffcdf9173884)

**Comment by samwhale** [comment]
aria-selected is not supported on checkbox inputs:
![image.png](https://images.zenhubusercontent.com/5b1032a1348b7966b43421a4/9033c6fc-b478-4513-a925-9b8b78712119)

**Comment by BrittanyIRL** [comment]
@samwhale can you point this?

**Comment by samwhale** [comment]
After some more investigation: the `listbox` role will not work with the tree structure because `option`s may not have nested focusable elements

Going back to the `tree` role

**Comment by samwhale** [comment]
To test:

1. Hitting `Tab` will no longer go through the options, instead it will move focus onto or away from the checkbox area
2. Tabbing into the checkbox area from the search box will do one of two things:
    - If there is a selected category then it will be focused
    - If there is not a selected category, then the first category will be focused
3. Tab into the checkbox list, change the focus with the arrow keys. Then tab back out and in. The checkbox that was focused before tabbing out should still receive focus.
4. Should be usable with a screen reader :)

**Comment by csossi** [comment]
Verified in QA
